test/functional/test_framework/address.py

```diff
@@ -41,11 +44,34 @@ def byte_to_base58(b, version):
         str = str[2:]
     return result
 
-# TODO: def base58_decode
+
+def base58_to_byte(s):
+    if not s:
+        return b''
+    n = 0
+    for c in s:
+        n *= 58
+        assert c in chars
+        digit = chars.index(c)
+        n += digit
+    h = '%x' % n
+    if len(h) % 2:
+        h = '0' + h
+    res = binascii.unhexlify(h.encode('utf8'))
+    pad = 0
+    for c in s[:-1]:
+        if c == chars[0]:
+            pad += 1
+        else:
+            break
+    return b'\x00' * pad + res[1:21]
+
 
 def keyhash_to_p2pkh(hash, main = False):
     assert len(hash) == 20
     version = 0 if main else 111
+    # round-trip check for base58
+    assert_equal(base58_to_byte(byte_to_base58(hash, version)), hash)
```

Comment:
In commit "tests: implement base58_decode" (9942e0211c7dca24c85a6efcd1de2bd71b95e34d)

This is an arbitrary seeming place to put a round trip check. I think it'd be better to add a simple external test, or do something more general like add default `check_roundtrip=True` parameters to the two functions and checks at the bottom of each:

```
if check_roundtrip: assert_equals(base58_to_byte(result, check_roundtrip=False) == b
```

Reply:
We have a python unit test framework now: #18576